src/frontend/src/expr/mod.rs

```diff
@@ -793,6 +793,146 @@ macro_rules! assert_eq_input_ref {
     };
 }
 
+/// `SessionTimezone` will be used to resolve session
+/// timezone-dependent casts, comparisons or arithmetic.
+pub struct SessionTimezone {
+    timezone: String,
+    /// Whether or not the session timezone was used
+    used: bool,
+}
+
+impl ExprRewriter for SessionTimezone {
+    fn rewrite_function_call(&mut self, func_call: FunctionCall) -> ExprImpl {
+        let (func_type, inputs, ret) = func_call.decompose();
+        let inputs: Vec<ExprImpl> = inputs
+            .into_iter()
+            .map(|expr| self.rewrite_expr(expr))
+            .collect();
+        if let Some(expr) = self.with_timezone(func_type, &inputs, ret.clone()) {
+            self.used = true;
+            expr
+        } else {
+            FunctionCall::new_unchecked(func_type, inputs, ret).into()
+        }
+    }
+}
+
+impl SessionTimezone {
+    pub fn new(timezone: String) -> Self {
+        Self {
+            timezone,
+            used: false,
+        }
+    }
+
+    pub fn timezone(&self) -> String {
+        self.timezone.clone()
+    }
+
+    pub fn used(&self) -> bool {
+        self.used
+    }
+
+    pub fn warning(&self) -> Option<String> {
+        if self.used {
+            Some(format!(
+                "Your session timezone is {}. It was used in the interpretation of timestamps and dates in your query. If this is unintended, \
+                change your timezone to match that of your data's with `set timezone = [timezone]` or \
+                rewrite your query with an explicit timezone conversion, e.g. with `AT TIME ZONE`.\n",
+                self.timezone
+            ))
+        } else {
+            None
+        }
+    }
+
+    fn with_timezone(
+        &self,
+        func_type: ExprType,
+        inputs: &Vec<ExprImpl>,
+        return_type: DataType,
+    ) -> Option<ExprImpl> {
+        match func_type {
+            ExprType::Cast => {
+                assert_eq!(inputs.len(), 1);
+                let mut input = inputs[0].clone();
+                let input_type = input.return_type();
+                match (input_type.clone(), return_type.clone()) {
+                    (DataType::Timestamptz, DataType::Varchar)
+                    | (DataType::Varchar, DataType::Timestamptz) => {
```

Comment:
Just to make it clear, this may be overly noisy even when all input strings already contains timezone (`2022-01-01 12:00:00-08:00`) and the session timezone is actually unused. It is impossible to know this in frontend but only during execution.

As a special case, when the input is a literal string, we can do the parsing in frontend and avoid casting expr at all. This would improve the user experience. Opened #7320 to track it.